Refactoring

lukdva-visma opened the pull request.

lukdva-visma committed:
Move classes to their rightful places/ rename accordingly

lukdva-visma committed:
Move business logic from controller to service
Move interaction with repository from controller to service
Remove service dependency from repository
Add repository dependency to service
Add updateMeeting method to meetingsService which replaces meeting with updated one and writes changes to repository

lukdva-visma committed:
Throw reasonable exception if jwt token fails to parse personId
Remove unused parts of code from MeetingController
Rename meetings to meetingsService in MeetingController
Move filtering logic meetingsService from controller
Use Stream for filtering meetings
Substitute filtering methods to Predicates
Use Spring beans for MeetingsService and MeetingsRepository
Autowire dependency injections

lukdva-visma committed:
Move @Getter @Setter annotations to class level
Inject objectMapper to MeetingsRepository
Use LocalDateTime class instead of deprecated Date class
Create MeetingsFilters class for easier parameters handling in controller
Fix Meeting tests

lukdva-visma committed:
Move @Getter @Setter annotations to class level
Extract .env variables with spring instead of dotenv
Convert JwtUtils to a spring bean
Remove dotenv dependency

lukdva-visma committed:
Resolve error: Plugin 'org.springframework.boot:spring-boot-maven-plugin:' not found

lauris94 commented on src/main/java/com/example/Meetings_API/Controllers/MeetingController.java:
```diff
@@ -1,119 +1,62 @@
 
 @RestController
 public class MeetingController {
+    @Autowired
```
Kodėl nenaudoji constructor injection'o?

lauris94 replied:
Dar nepataisyta

lauris94 commented on src/main/java/com/example/Meetings_API/Controllers/MeetingController.java:
```diff
@@ -1,119 +1,62 @@
-        meetings.addMeeting(meeting);
-        addAttendee(meeting.getId(), meeting.getResponsiblePerson());
-        meetingsService.writeMeetings();
+        meetingsService.addAttendeeToMeeting(meeting, meeting.getResponsiblePerson());
```
servisą callink 1 kartą

lauris94 commented on src/main/java/com/example/Meetings_API/Controllers/MeetingController.java:
```diff
@@ -1,119 +1,62 @@
-        }
+        Meeting meeting = meetingsService.getMeeting(id);
         String personId = jwtUtils.getPersonIdFromToken(token);
         if(meeting.getResponsiblePerson().getId().equals(personId))
```
šita logika turėtų būt servise

lauris94 commented on src/main/java/com/example/Meetings_API/Controllers/MeetingController.java:
```diff
@@ -1,119 +1,62 @@
-        meetingsService.writeMeetings();
+        Meeting meeting = meetingsService.getMeeting(id);
+        Meeting updatedMeeting = meetingsService.addAttendeeToMeeting(meeting, person);
+        meetingsService.updateMeeting(updatedMeeting);
```
vienas callsas turėtų būt į servisą, vietoj 3

lauris94 commented on src/main/java/com/example/Meetings_API/Controllers/MeetingController.java:
```diff
@@ -1,119 +1,62 @@
-                throw new RuntimeException(e);
-            }
-        return filteredMeetings.getMeetings();
+    public List<Meeting> getMeetings(MeetingsFilters filters) {
```
MeetingFilter gal geriau?

lauris94 commented on src/main/java/com/example/Meetings_API/Controllers/MeetingController.java:
```diff
@@ -1,119 +1,62 @@
-        meetingsService.readMeetings();
-    }
     @PostMapping("/meetings")
     public Meeting createMeeting(@RequestBody Meeting meeting) {
```
Sukurk atskirą objektą request bodžiui

lukdva-visma committed:
Create Meeting and Person DTO classes and assemblers for mapping DTO to Model objects
Make only single call to Meeting service from controller methods

lukdva-visma committed:
Fix Meeting service unit tests
Fix isAttendeeAvailable method
Create new File instance in Configuration file as a bean

lauris94 commented on pom.xml:
```diff
@@ -30,18 +30,13 @@
 			<groupId>org.springframework.boot</groupId>
 			<artifactId>spring-boot-starter-test</artifactId>
-			<scope>test</scope>
+			<version>2.7.5</version>
```
kodėl kitokia versija nei spring boot parent?

lauris94 replied:
kodėl kitokia versija nei spring boot parent?

lauris94 commented on src/main/java/com/example/Meetings_API/Assemblers/MeetingAssembler.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.stereotype.Component;
+
+import java.time.LocalDateTime;
+@Component
```
Mapper klasės neturėtų būt beansai, nes neturėtų reikėt injectint kitų beansų ir pnš. Galima turėt tiesiog static metodus mapperiuose

lauris94 commented on src/main/java/com/example/Meetings_API/Assemblers/MeetingAssembler.java:
```diff
@@ -0,0 +1,22 @@
+@Component
+public class MeetingAssembler {
+    public Meeting mapMeeting(MeetingDTO meetingDto)
+    {
```
Formatavimas neatitinka javos taisyklių, ctrl + alt + l turėtų automatiškai suformatuot

lauris94 commented on src/main/java/com/example/Meetings_API/Models/MeetingDTO.java:
```diff
@@ -0,0 +1,19 @@
+@Getter @Setter
+public class MeetingDTO {
+    private String name;
+    private Person responsiblePerson;
```
turėtų būt PersonDTO

lauris94 commented on src/test/java/com/example/Meetings_API/Services/MeetingServiceUnitTest.java:
```diff
@@ -0,0 +1,298 @@
+    void verifyAddMeetingMethodAddsMeetingAndCallsRepositoryWithUpdated() {
+        Meeting[] expectedMeetingsList = {meeting};
+        meetingsService.addMeeting(meeting);
+        assertArrayEquals(meetingsService.getFilteredMeetings(new MeetingsFilters()).toArray(), expectedMeetingsList);
```
Tokiu atveju gaunasi, kad testuoji 2 serviso metodus, tai nevisai unit testas tada. Siūlau vietoj tokio asserto naudoti ArgumentCaptor ir paasertinti kas buvo paduota į repository.writeMeetings

lukdva-visma replied:
O kam reikia ArgumentCaptor? 
_verify(repository).writeMeetings(Arrays.asList(expectedMeetingsList));_ patikrina kad buvo call'intas _writeMeetings_ su _expectedMeetingList_.
Ar better practice butu naudoti ArgumentCaptor, ir issitraukus ta list'a patikrinti kad turi viena elementa ir jis equals su pries tai nauduotu _meeting_?

lauris94 commented on src/test/java/com/example/Meetings_API/Services/MeetingServiceUnitTest.java:
```diff
@@ -0,0 +1,298 @@
+    @Test
+    @DisplayName("Add meeting method adds meeting to list and calls repository with updated list")
+    void verifyAddMeetingMethodAddsMeetingAndCallsRepositoryWithUpdated() {
+        Meeting[] expectedMeetingsList = {meeting};
```
gali naudot listą vietoj array, galėtum rašyt pvz List.of(..)

lauris94 commented on src/test/java/com/example/Meetings_API/Services/MeetingServiceUnitTest.java:
```diff
@@ -0,0 +1,298 @@
+    @Test
+    @DisplayName("getMeeting returns existing meeting")
+    void getExistingMeetingReturnsAddedMeeting() {
+        meetingsService.addMeeting(meeting);
```
Irgi 2 metodus testuoji, naudok mockus ant repository callsų

lukdva-visma replied:
Testas skirtas getMeeting metodui. Su addMeeting susikuriu, kad nebūtų empty list'as.

lauris94 commented on src/main/java/com/example/Meetings_API/Services/MeetingsService.java:
```diff
@@ -1,37 +1,144 @@
+    }
+    public void deleteMeeting(String meetingId, String personId)
+    {
+        Meeting meeting = getMeeting(meetingId);
```
nemappink į meeting objektą be reikalo, juk gali removint meetingą pagal id

lukdva-visma committed:
Move DTO classes to separate package
Reformat all classes

lauris94 commented on src/main/java/com/example/Meetings_API/Services/MeetingsService.java:
```diff
@@ -1,37 +1,144 @@
+    }
+    public Meeting getMeeting(String id)
+    {
+        for (Meeting meeting: meetings
```
panaudok stream

lauris94 commented on src/test/java/com/example/Meetings_API/Services/MeetingServiceUnitTest.java:
```diff
@@ -0,0 +1,310 @@
+    @Test
+    void checkMeetingIsRemovedSuccessfully() {
+        meetingsService.addMeeting(meeting);
+        reset(repository); // test does not care calls to mock before this statement as it is a setup
```
o kur repository.readMeetings() mockas?

lukdva-visma replied:
readMeetings() atliekamas tik MeetingsService konstruktoriuje. Šitam teste tai nepachekinsiu. Ar reiktu konstruktoriui testu parašyt?
Ar čia apie kažką kito klausei?

lukdva-visma committed:
Use Constuctor based dependency injection
Get file path of repository file from application.properties
Use PersonDTO class instead of Person in MeetingDTO
Make adjustments to MeetingService tests
Rename MeetingsFilters to MeetingFilter
Use ConfigProperties class to inject properties

lukdva-visma committed:
Add MeetingBuilder and PersonBuilder classes for easier data objects creation
Add TestUtils class to put test helper methods there
Refactor tests to use Builder classes
Use assertIterableEquals instead of assertArrayEquals
Use Stubs for initial test data state instead of actual class methods to add data
Add tests for not yet covered MeetingsService class methods

lukdva-visma committed:
Add curly braces for if statements which don't have them

lukdva-visma committed:
Use MeetingResponseDTO for response bodies where applicable

lauris94 commented on src/main/java/com/example/Meetings_API/Controllers/MeetingController.java:
```diff
@@ -1,119 +1,72 @@
 package com.example.Meetings_API.Controllers;
```
package name'ai turi būti lowercase

lauris94 commented on src/main/java/com/example/Meetings_API/Controllers/MeetingController.java:
```diff
@@ -1,119 +1,72 @@
-        meetingsService.readMeetings();
+    JwtUtils jwtUtils;
+
+    ConfigProperties config;
```
nenaudojamas fieldas

lauris94 commented on src/main/java/com/example/Meetings_API/DTOs/MeetingResponseDTO.java:
```diff
@@ -0,0 +1,40 @@
+    public MeetingResponseDTO() {
+    }
+
+    public MeetingResponseDTO(String id, String name, PersonDTO responsiblePerson, String description, Category category, Type type, LocalDateTime startDate, LocalDateTime endDate, List<Attendee> attendees) {
```
Nedaryk tokio didelio konstruktoriaus, paprastai konstruktioriuj includina tik privalomus fieldus. Šiuo atveju galima settint naudojant setterius

lauris94 commented on src/main/java/com/example/Meetings_API/Services/MeetingsService.java:
```diff
@@ -1,37 +1,163 @@
-        } catch (IOException e) {
-            e.printStackTrace();
+    public List<Meeting> listOfMeetingsPersonIsIn(Person person) {
+        List<Meeting> list = new ArrayList<>();
```
panaudok java stream

lukdva-visma committed:
Replace @InjectMocks annotation with meetingService creation at @BeforeEach method
Adjust stubbing of test data
Replace constructor with @RequiredArgsConstructor in MeetingController
Remove unnecessary variable declarations in MeetingController

lukdva-visma committed:
Add logging to ExceptionHandler class
Retrieve response error messages from messages.properties file instead of hardcoding

lukdva-visma committed:
Rename packages to start with lower case
Use stream instead of for loops
Remove constructor with too many parameters for MeetingResponseDTO
